test/UnitTestHelpers.hs

```diff
@@ -0,0 +1,352 @@
+import           Game.LambdaHack.Content.TileKind
+import qualified Game.LambdaHack.Core.Dice as Dice
+
+import qualified Game.LambdaHack.Definition.Ability as Ability
+import           Game.LambdaHack.Definition.Color
+import           Game.LambdaHack.Definition.Flavour
+
+import Content.FactionKind
+
+import Game.LambdaHack.Common.Misc
+
+
+-- Read UI requests from the client and send them to the frontend,
+fchanFrontendStub :: ChanFrontend
+fchanFrontendStub =
+  ChanFrontend $ \case
+    FrontFrame _ -> putStr "FrontFrame"
+    FrontDelay _ -> putStr "FrontDelay"
+    FrontKey _ _ -> return KMP { kmpKeyMod=K.escKM, kmpPointer=PointUI 0 0 }
+    FrontPressed -> return False
+    FrontDiscardKey -> putStr "FrontDiscardKey"
+    FrontResetKeys -> putStr "FrontResetKeys"
+    FrontShutdown -> putStr "FrontShutdown"
+    FrontPrintScreen -> putStr "FrontPrintScreen"
+
+data CliState = CliState
+  { cliState   :: State            -- ^ current global state
+  , cliClient  :: StateClient      -- ^ current client state
+  , cliSession :: Maybe SessionUI  -- ^ UI state, empty for AI clients
+  -- , cliDict    :: ChanServer       -- ^ this client connection information
+  -- , cliToSave  :: Save.ChanSave (StateClient, Maybe SessionUI)
+  }
+
+
+stubUIOptions :: UIOptions
+stubUIOptions = UIOptions
+  { uCommands = []
+  , uHeroNames = []
+  , uVi = False
+  , uLeftHand = False
+  , uChosenFontset = ""
+  , uAllFontsScale = 0.0
+  , uFullscreenMode = NotFullscreen      
+  , uhpWarningPercent = 0
+  , uMsgWrapColumn = 0
+  , uHistoryMax = 0
+  , uMaxFps = 0.0
+  , uNoAnim = False
+  , uOverrideCmdline = []
+  , uFonts = []
+  , uFontsets = []
+  , uMessageColors = []
+  }
+
+stubClientOptions :: ClientOptions
+stubClientOptions = defClientOptions 
+  { schosenFontset = Just "snoopy"
+  , sfontsets = [("snoopy",FontSet {fontMapScalable="scalable",fontMapBitmap="bitmap",fontPropRegular="propRegular",fontPropBold="propBold",fontMono="mono"})]
+  }
+
+testLevelDimension :: Int
+testLevelDimension = 3
+
+-- using different arbitrary numbers for these so if tests fail to missing keys we'll have more of a clue
+testLevelId :: LevelId 
+testLevelId = toEnum 111
+
+testActorId :: ActorId
+testActorId = toEnum 112
+
+testItemId :: ItemId 
+testItemId = toEnum 113
+
+testFactionId :: FactionId 
+testFactionId = toEnum 114
+
+
+
+testArea :: Area
+testArea = fromJust(toArea (0, 0, 0, 0))
+
+stubLevel :: Level
+stubLevel = Level
+  { lkind = toEnum 0
+  , ldepth = Dice.AbsDepth 1
+  , lfloor = EM.empty
+  , lembed = EM.empty
+  , lbig = EM.empty
+  , lproj = EM.empty
+  , ltile = unknownTileMap testArea unknownId testLevelDimension testLevelDimension
+  , lentry = EM.empty
+  , larea = trivialArea (Point 0 0)
+  , lsmell = EM.empty
+  , lstair = ([],[])
+  , lescape = []
+  , lseen = 0
+  , lexpl = 0
+  , ltime = timeZero
+  , lnight = False
+  }
+
+
+testFaction :: Faction
+testFaction =
+  Faction
+    { gkind = head content 
```

Comment:
This is fine, but it's not a stub, but real content taken from `Content.FactionKind`.

Reply:
And the only way to get a stub is to laboriously fill a FactionKind record with dummy values.